urbanairship-react-native/js/AttributeEditor.ts

```diff
@@ -0,0 +1,100 @@
+/* Copyright Airship and Contributors */
+
+'use strict';
+
+/**
+ * Attribute operation
+ * @hidden
+ */
+export interface AttributeOperation {
+  /**
+   * The operation name.
+   */
+  action: string;
+  /**
+   * The attribute key.
+   */
+  key: string;
+  /**
+   * The attribute value, if avaialble.
+   */
+  value?: string | number | Date;
+  /**
+   * The attribute type, if available.
+   */
+  type?: "string" | "number" | "date";
+}
+
+/**
+ * Editor for attributes.
+ */
+export class AttributeEditor {
+  onApply: (operations: AttributeOperation[]) => void;
+  operations: AttributeOperation[];
+
+  /**
+   * AttributeEditor constructor
+   *
+   * @hidden
+   * @param onApply The apply function
+   */
+  constructor(onApply: (operations: AttributeOperation[]) => void) {
+    this.onApply = onApply;
+    this.operations = [];
+  }
+
+  /**
+   * Adds an attribute.
+   *
+   * @param value The attribute value.
+   * @param name The attribute name.
+   * @return The attribute editor instance.
+   */
+  setAttribute(name: string, value: string | number | boolean | Date): AttributeEditor {
+    var attributeValue: string | number | Date;
+    var attributeType: "string" | "number" | "date";
+
+    if (typeof value != "boolean") {
+      attributeValue = value;
+      if (typeof value === "string") {
+        attributeType = "string";
+      } else if (typeof attributeValue === "number") {
+        attributeType = "number";
+      } else if (value instanceof Date) {
+        // JavaScript's date type doesn't pass through the JS to native bridge.
+        // Dates are instead serialized as milliseconds since epoch.
+        attributeType = "date";
+        attributeValue = value.getTime();
+      } else {
+        throw "Unsupported attribute type: " + typeof attributeValue;
+      }
+    } else {
+      // No boolean attribute type. Convert value to string.
+      attributeValue = value.toString();
+      attributeType = "string";
+    }
+
+    const operation = { "action": "set", "value": attributeValue, "key": name, type: attributeType};
```

Comment:
AttributeOperation is an interface, not a class. Interfaces in TypeScript are similar to those in Go – any object with those keys and value types already implements it implicitly. This also helps for compatibility with plain JavaScript since you can just keep passing an object/map like before and not have to worry about constructors/classes/etc.